<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>12 Genome Environment Association | Semester.knit</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="12 Genome Environment Association | Semester.knit" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="12 Genome Environment Association | Semester.knit" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="week-10-genome-wide-association-study-gwas.html"/>

<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Marine Genomics Semester</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Marine Genomics Semester edition</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#cloud-computing-versus-individual-laptops"><i class="fa fa-check"></i><b>1.1</b> Cloud computing versus individual laptops</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#assessment"><i class="fa fa-check"></i><b>1.2</b> Assessment</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#short-recorded-demonstrations"><i class="fa fa-check"></i><b>1.3</b> Short recorded demonstrations</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#final-demonstration"><i class="fa fa-check"></i><b>1.4</b> Final demonstration</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="week-1--welcome.html"><a href="week-1--welcome.html"><i class="fa fa-check"></i><b>2</b> Week 1- Welcome!</a>
<ul>
<li class="chapter" data-level="2.1" data-path="week-1--welcome.html"><a href="week-1--welcome.html#introduction-to-shell-computing-via-the-data-carpentry-tutorial"><i class="fa fa-check"></i><b>2.1</b> Introduction to shell computing via the data carpentry tutorial</a></li>
<li class="chapter" data-level="2.2" data-path="week-1--welcome.html"><a href="week-1--welcome.html#how-to-access-the-shell"><i class="fa fa-check"></i><b>2.2</b> How to access the shell</a></li>
<li class="chapter" data-level="2.3" data-path="week-1--welcome.html"><a href="week-1--welcome.html#best-practices-for-jetstream"><i class="fa fa-check"></i><b>2.3</b> Best practices for Jetstream</a></li>
<li class="chapter" data-level="2.4" data-path="week-1--welcome.html"><a href="week-1--welcome.html#week-1-objectives"><i class="fa fa-check"></i><b>2.4</b> Week 1 Objectives</a></li>
<li class="chapter" data-level="2.5" data-path="week-1--welcome.html"><a href="week-1--welcome.html#navigating-your-file-system"><i class="fa fa-check"></i><b>2.5</b> Navigating your file system</a></li>
<li class="chapter" data-level="2.6" data-path="week-1--welcome.html"><a href="week-1--welcome.html#shortcut-tab-completion"><i class="fa fa-check"></i><b>2.6</b> Shortcut: Tab Completion</a></li>
<li class="chapter" data-level="2.7" data-path="week-1--welcome.html"><a href="week-1--welcome.html#summary-key-points"><i class="fa fa-check"></i><b>2.7</b> Summary &amp; Key Points</a></li>
<li class="chapter" data-level="2.8" data-path="week-1--welcome.html"><a href="week-1--welcome.html#navigating-files-and-directories"><i class="fa fa-check"></i><b>2.8</b> Navigating Files and Directories</a></li>
<li class="chapter" data-level="2.9" data-path="week-1--welcome.html"><a href="week-1--welcome.html#moving-around-the-file-system"><i class="fa fa-check"></i><b>2.9</b> Moving around the file system</a></li>
<li class="chapter" data-level="2.10" data-path="week-1--welcome.html"><a href="week-1--welcome.html#navigating-practice-and-finding-hidden-directories"><i class="fa fa-check"></i><b>2.10</b> Navigating practice and finding hidden directories</a></li>
<li class="chapter" data-level="2.11" data-path="week-1--welcome.html"><a href="week-1--welcome.html#examining-the-contents-of-other-directories"><i class="fa fa-check"></i><b>2.11</b> Examining the contents of other directories</a></li>
<li class="chapter" data-level="2.12" data-path="week-1--welcome.html"><a href="week-1--welcome.html#full-vs-relative-paths"><i class="fa fa-check"></i><b>2.12</b> Full vs Relative Paths</a></li>
<li class="chapter" data-level="2.13" data-path="week-1--welcome.html"><a href="week-1--welcome.html#navigational-shortcuts"><i class="fa fa-check"></i><b>2.13</b> Navigational shortcuts</a></li>
<li class="chapter" data-level="2.14" data-path="week-1--welcome.html"><a href="week-1--welcome.html#key-points"><i class="fa fa-check"></i><b>2.14</b> Key Points</a></li>
<li class="chapter" data-level="2.15" data-path="week-1--welcome.html"><a href="week-1--welcome.html#creature-of-the-week"><i class="fa fa-check"></i><b>2.15</b> Creature of the Week!</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html"><i class="fa fa-check"></i><b>3</b> Week 2: Working With Files</a>
<ul>
<li class="chapter" data-level="3.1" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#our-data-set-fastq-files"><i class="fa fa-check"></i><b>3.1</b> Our data set: FASTQ files</a></li>
<li class="chapter" data-level="3.2" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#wildcards"><i class="fa fa-check"></i><b>3.2</b> Wildcards</a></li>
<li class="chapter" data-level="3.3" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise"><i class="fa fa-check"></i><b>3.3</b> Exercise</a></li>
<li class="chapter" data-level="3.4" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#using-echo"><i class="fa fa-check"></i><b>3.4</b> Using echo</a></li>
<li class="chapter" data-level="3.5" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#command-history"><i class="fa fa-check"></i><b>3.5</b> Command History</a></li>
<li class="chapter" data-level="3.6" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#examining-files"><i class="fa fa-check"></i><b>3.6</b> Examining Files</a></li>
<li class="chapter" data-level="3.7" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-1"><i class="fa fa-check"></i><b>3.7</b> Exercise</a></li>
<li class="chapter" data-level="3.8" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-2"><i class="fa fa-check"></i><b>3.8</b> Exercise</a></li>
<li class="chapter" data-level="3.9" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#details-on-the-fastq-format"><i class="fa fa-check"></i><b>3.9</b> Details on the FASTQ format</a></li>
<li class="chapter" data-level="3.10" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#creating-moving-copying-and-removing"><i class="fa fa-check"></i><b>3.10</b> Creating, moving, copying, and removing</a>
<ul>
<li class="chapter" data-level="3.10.1" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#copying-files"><i class="fa fa-check"></i><b>3.10.1</b> Copying Files</a></li>
<li class="chapter" data-level="3.10.2" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#creating-directories"><i class="fa fa-check"></i><b>3.10.2</b> Creating Directories</a></li>
<li class="chapter" data-level="3.10.3" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#moving-renaming"><i class="fa fa-check"></i><b>3.10.3</b> Moving / Renaming</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#file-permissions"><i class="fa fa-check"></i><b>3.11</b> File Permissions</a></li>
<li class="chapter" data-level="3.12" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#removing"><i class="fa fa-check"></i><b>3.12</b> Removing</a></li>
<li class="chapter" data-level="3.13" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-3"><i class="fa fa-check"></i><b>3.13</b> Exercise</a></li>
<li class="chapter" data-level="3.14" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#redirections"><i class="fa fa-check"></i><b>3.14</b> Redirections</a></li>
<li class="chapter" data-level="3.15" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#searching-files"><i class="fa fa-check"></i><b>3.15</b> Searching files</a></li>
<li class="chapter" data-level="3.16" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-4"><i class="fa fa-check"></i><b>3.16</b> Exercise</a></li>
<li class="chapter" data-level="3.17" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#redirecting-output"><i class="fa fa-check"></i><b>3.17</b> Redirecting output</a>
<ul>
<li class="chapter" data-level="3.17.1" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-5"><i class="fa fa-check"></i><b>3.17.1</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.18" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#writing-for-loops"><i class="fa fa-check"></i><b>3.18</b> Writing for loops</a></li>
<li class="chapter" data-level="3.19" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#using-basename-in-for-loops"><i class="fa fa-check"></i><b>3.19</b> Using Basename in for loops</a></li>
<li class="chapter" data-level="3.20" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-6"><i class="fa fa-check"></i><b>3.20</b> Exercise</a></li>
<li class="chapter" data-level="3.21" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-7"><i class="fa fa-check"></i><b>3.21</b> Exercise</a></li>
<li class="chapter" data-level="3.22" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#writing-scripts-and-working-with-data"><i class="fa fa-check"></i><b>3.22</b> Writing Scripts and Working with Data</a></li>
<li class="chapter" data-level="3.23" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#writing-files"><i class="fa fa-check"></i><b>3.23</b> Writing files</a></li>
<li class="chapter" data-level="3.24" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-8"><i class="fa fa-check"></i><b>3.24</b> Exercise</a></li>
<li class="chapter" data-level="3.25" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#writing-scripts"><i class="fa fa-check"></i><b>3.25</b> Writing scripts</a></li>
<li class="chapter" data-level="3.26" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#exercise-9"><i class="fa fa-check"></i><b>3.26</b> Exercise</a></li>
<li class="chapter" data-level="3.27" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#making-the-script-into-a-program"><i class="fa fa-check"></i><b>3.27</b> Making the script into a program</a></li>
<li class="chapter" data-level="3.28" data-path="week-2-working-with-files.html"><a href="week-2-working-with-files.html#creature-of-the-week-1"><i class="fa fa-check"></i><b>3.28</b> Creature of the Week!</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="week3-awk-and-how-to-download-files.html"><a href="week3-awk-and-how-to-download-files.html"><i class="fa fa-check"></i><b>4</b> Week3: Awk and how to download files</a>
<ul>
<li class="chapter" data-level="4.1" data-path="week3-awk-and-how-to-download-files.html"><a href="week3-awk-and-how-to-download-files.html#awk"><i class="fa fa-check"></i><b>4.1</b> AWK</a></li>
<li class="chapter" data-level="4.2" data-path="week3-awk-and-how-to-download-files.html"><a href="week3-awk-and-how-to-download-files.html#moving-and-downloading-data"><i class="fa fa-check"></i><b>4.2</b> Moving and Downloading Data</a></li>
<li class="chapter" data-level="4.3" data-path="week3-awk-and-how-to-download-files.html"><a href="week3-awk-and-how-to-download-files.html#getting-data-from-the-cloud"><i class="fa fa-check"></i><b>4.3</b> Getting data from the cloud</a></li>
<li class="chapter" data-level="4.4" data-path="week3-awk-and-how-to-download-files.html"><a href="week3-awk-and-how-to-download-files.html#downloading-files-from-github"><i class="fa fa-check"></i><b>4.4</b> Downloading files from Github</a></li>
<li class="chapter" data-level="4.5" data-path="week3-awk-and-how-to-download-files.html"><a href="week3-awk-and-how-to-download-files.html#final-bash-wrap-up"><i class="fa fa-check"></i><b>4.5</b> Final bash wrap-up</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html"><i class="fa fa-check"></i><b>5</b> Week 4- What is a Genetic Variant?</a>
<ul>
<li class="chapter" data-level="5.1" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html#to-get-started-lets-download-the-data-and-install-a-few-programs"><i class="fa fa-check"></i><b>5.1</b> To get started lets download the data and install a few programs</a></li>
<li class="chapter" data-level="5.2" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html#raw-read-quality-control"><i class="fa fa-check"></i><b>5.2</b> Raw read quality control</a></li>
<li class="chapter" data-level="5.3" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html#trimming-to-remove-adapters"><i class="fa fa-check"></i><b>5.3</b> Trimming to remove adapters</a></li>
<li class="chapter" data-level="5.4" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html#building-an-index-of-our-genome"><i class="fa fa-check"></i><b>5.4</b> Building an index of our genome</a></li>
<li class="chapter" data-level="5.5" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html#map-reads-to-the-genome"><i class="fa fa-check"></i><b>5.5</b> Map reads to the genome</a></li>
<li class="chapter" data-level="5.6" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html#sam-to-bam-file-conversion"><i class="fa fa-check"></i><b>5.6</b> sam to bam file conversion</a></li>
<li class="chapter" data-level="5.7" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html#genotype-likelihoods"><i class="fa fa-check"></i><b>5.7</b> Genotype likelihoods</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="week-4--what-is-a-genetic-variant.html"><a href="week-4--what-is-a-genetic-variant.html#suggested-exercises"><i class="fa fa-check"></i><b>5.7.1</b> Suggested Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html"><i class="fa fa-check"></i><b>6</b> Week 5- The R environment</a>
<ul>
<li class="chapter" data-level="6.1" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#lesson-1-orientation-to-r"><i class="fa fa-check"></i><b>6.1</b> Lesson 1: Orientation to R</a></li>
<li class="chapter" data-level="6.2" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#manipulating-a-vector-object"><i class="fa fa-check"></i><b>6.2</b> Manipulating a vector object</a></li>
<li class="chapter" data-level="6.3" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#operations-act-on-each-element-of-a-vector"><i class="fa fa-check"></i><b>6.3</b> Operations act on each element of a vector:</a></li>
<li class="chapter" data-level="6.4" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#operations-can-also-work-with-two-vectors"><i class="fa fa-check"></i><b>6.4</b> Operations can also work with two vectors:</a></li>
<li class="chapter" data-level="6.5" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#a-few-tips-below-for-working-with-objects"><i class="fa fa-check"></i><b>6.5</b> A few tips below for working with objects:</a></li>
<li class="chapter" data-level="6.6" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#exercise-1.1"><i class="fa fa-check"></i><b>6.6</b> EXERCISE 1.1</a></li>
<li class="chapter" data-level="6.7" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#characterizing-a-dataframe"><i class="fa fa-check"></i><b>6.7</b> 1.2 Characterizing a dataframe</a></li>
<li class="chapter" data-level="6.8" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#how-to-access-parts-of-the-data"><i class="fa fa-check"></i><b>6.8</b> How to access parts of the data:</a></li>
<li class="chapter" data-level="6.9" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#data-manipulation"><i class="fa fa-check"></i><b>6.9</b> Data Manipulation</a></li>
<li class="chapter" data-level="6.10" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#exercise-1.2"><i class="fa fa-check"></i><b>6.10</b> EXERCISE 1.2</a></li>
<li class="chapter" data-level="6.11" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#subsetting-datasets-logicals"><i class="fa fa-check"></i><b>6.11</b> 1.3 Subsetting datasets &amp; logicals</a></li>
<li class="chapter" data-level="6.12" data-path="week-5--the-r-environment.html"><a href="week-5--the-r-environment.html#exercise-1.3-indexing-by-logical-statements"><i class="fa fa-check"></i><b>6.12</b> EXERCISE 1.3 indexing by logical statements</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="week-6--r-continued.html"><a href="week-6--r-continued.html"><i class="fa fa-check"></i><b>7</b> Week 6- R Continued</a>
<ul>
<li class="chapter" data-level="7.1" data-path="week-6--r-continued.html"><a href="week-6--r-continued.html#plotting"><i class="fa fa-check"></i><b>7.1</b> 2.1 Plotting</a></li>
<li class="chapter" data-level="7.2" data-path="week-6--r-continued.html"><a href="week-6--r-continued.html#scatterplots"><i class="fa fa-check"></i><b>7.2</b> Scatterplots</a></li>
<li class="chapter" data-level="7.3" data-path="week-6--r-continued.html"><a href="week-6--r-continued.html#customizing-your-plot"><i class="fa fa-check"></i><b>7.3</b> Customizing your plot</a></li>
<li class="chapter" data-level="7.4" data-path="week-6--r-continued.html"><a href="week-6--r-continued.html#exercise-2.1"><i class="fa fa-check"></i><b>7.4</b> Exercise 2.1</a></li>
<li class="chapter" data-level="7.5" data-path="week-6--r-continued.html"><a href="week-6--r-continued.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>7.5</b> 2.2 plotting with ggplot2</a></li>
<li class="chapter" data-level="7.6" data-path="week-6--r-continued.html"><a href="week-6--r-continued.html#exercise-2.2-plotting-in-ggplot2"><i class="fa fa-check"></i><b>7.6</b> Exercise 2.2 Plotting in ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="week-7-for-loops-and-the-apply-family-of-functions.html"><a href="week-7-for-loops-and-the-apply-family-of-functions.html"><i class="fa fa-check"></i><b>8</b> Week 7 For loops and the apply family of functions</a>
<ul>
<li class="chapter" data-level="8.0.1" data-path="week-7-for-loops-and-the-apply-family-of-functions.html"><a href="week-7-for-loops-and-the-apply-family-of-functions.html#apply"><i class="fa fa-check"></i><b>8.0.1</b> apply</a></li>
<li class="chapter" data-level="8.0.2" data-path="week-7-for-loops-and-the-apply-family-of-functions.html"><a href="week-7-for-loops-and-the-apply-family-of-functions.html#lapply-list-apply"><i class="fa fa-check"></i><b>8.0.2</b> lapply – “list” apply</a></li>
<li class="chapter" data-level="8.0.3" data-path="week-7-for-loops-and-the-apply-family-of-functions.html"><a href="week-7-for-loops-and-the-apply-family-of-functions.html#tapply---per-type-apply"><i class="fa fa-check"></i><b>8.0.3</b> tapply - “per Type” apply</a></li>
<li class="chapter" data-level="8.1" data-path="week-7-for-loops-and-the-apply-family-of-functions.html"><a href="week-7-for-loops-and-the-apply-family-of-functions.html#exercise-2.3-apply-and-tapply"><i class="fa fa-check"></i><b>8.1</b> Exercise 2.3 apply and tapply</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="week-8--principle-component-analyses.html"><a href="week-8--principle-component-analyses.html"><i class="fa fa-check"></i><b>9</b> Week 8- Principle Component Analyses</a>
<ul>
<li class="chapter" data-level="9.1" data-path="week-8--principle-component-analyses.html"><a href="week-8--principle-component-analyses.html#download-the-data"><i class="fa fa-check"></i><b>9.1</b> Download the data</a></li>
<li class="chapter" data-level="9.2" data-path="week-8--principle-component-analyses.html"><a href="week-8--principle-component-analyses.html#installing-programs"><i class="fa fa-check"></i><b>9.2</b> Installing programs</a></li>
<li class="chapter" data-level="9.3" data-path="week-8--principle-component-analyses.html"><a href="week-8--principle-component-analyses.html#run-pcangsd-on-our-data"><i class="fa fa-check"></i><b>9.3</b> Run pcangsd on our data</a></li>
<li class="chapter" data-level="9.4" data-path="week-8--principle-component-analyses.html"><a href="week-8--principle-component-analyses.html#setting-up-a-new-project-in-r"><i class="fa fa-check"></i><b>9.4</b> Setting up a new project in R</a></li>
<li class="chapter" data-level="9.5" data-path="week-8--principle-component-analyses.html"><a href="week-8--principle-component-analyses.html#reading-data-into-r"><i class="fa fa-check"></i><b>9.5</b> Reading data into R</a></li>
<li class="chapter" data-level="9.6" data-path="week-8--principle-component-analyses.html"><a href="week-8--principle-component-analyses.html#exercises"><i class="fa fa-check"></i><b>9.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="week-9--fst-and-outlier-analysis.html"><a href="week-9--fst-and-outlier-analysis.html"><i class="fa fa-check"></i><b>10</b> Week 9- F<sub>st</sub> and outlier analysis</a>
<ul>
<li class="chapter" data-level="10.1" data-path="week-9--fst-and-outlier-analysis.html"><a href="week-9--fst-and-outlier-analysis.html#download-the-data-1"><i class="fa fa-check"></i><b>10.1</b> Download the data</a></li>
<li class="chapter" data-level="10.2" data-path="week-9--fst-and-outlier-analysis.html"><a href="week-9--fst-and-outlier-analysis.html#getting-r-set-up"><i class="fa fa-check"></i><b>10.2</b> Getting R set up</a></li>
<li class="chapter" data-level="10.3" data-path="week-9--fst-and-outlier-analysis.html"><a href="week-9--fst-and-outlier-analysis.html#finding-outliers-using-pcadapt"><i class="fa fa-check"></i><b>10.3</b> Finding outliers using pcadapt</a></li>
<li class="chapter" data-level="10.4" data-path="week-9--fst-and-outlier-analysis.html"><a href="week-9--fst-and-outlier-analysis.html#pcadapt-exercises"><i class="fa fa-check"></i><b>10.4</b> pcadapt Exercises</a></li>
<li class="chapter" data-level="10.5" data-path="week-9--fst-and-outlier-analysis.html"><a href="week-9--fst-and-outlier-analysis.html#using-fst-to-find-outliers"><i class="fa fa-check"></i><b>10.5</b> Using F<sub>st</sub> to find outliers</a></li>
<li class="chapter" data-level="10.6" data-path="week-9--fst-and-outlier-analysis.html"><a href="week-9--fst-and-outlier-analysis.html#outflank-practice"><i class="fa fa-check"></i><b>10.6</b> OutFLANK Practice</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>11</b> Week 10: Genome wide association study (GWAS)</a>
<ul>
<li class="chapter" data-level="11.1" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#load-in-the-data"><i class="fa fa-check"></i><b>11.1</b> load in the data</a></li>
<li class="chapter" data-level="11.2" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#install-packages-in-r"><i class="fa fa-check"></i><b>11.2</b> install packages in R</a></li>
<li class="chapter" data-level="11.3" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#install-angsd-and-pcangsd-again"><i class="fa fa-check"></i><b>11.3</b> install angsd and pcangsd (again?)</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#install-angsd"><i class="fa fa-check"></i><b>11.3.1</b> install angsd</a></li>
<li class="chapter" data-level="11.3.2" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#for-pcangsd"><i class="fa fa-check"></i><b>11.3.2</b> for pcangsd</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#the-data"><i class="fa fa-check"></i><b>11.4</b> The data</a></li>
<li class="chapter" data-level="11.5" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#take-beagle-file-and-generate-lrt-file"><i class="fa fa-check"></i><b>11.5</b> take Beagle file and generate lrt file</a></li>
<li class="chapter" data-level="11.6" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#take-lrt-file-and-make-a-manhattan-plot"><i class="fa fa-check"></i><b>11.6</b> take lrt file and make a manhattan plot</a></li>
<li class="chapter" data-level="11.7" data-path="week-10-genome-wide-association-study-gwas.html"><a href="week-10-genome-wide-association-study-gwas.html#exercises-1"><i class="fa fa-check"></i><b>11.7</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="genome-environment-association.html"><a href="genome-environment-association.html"><i class="fa fa-check"></i><b>12</b> Genome Environment Association</a>
<ul>
<li class="chapter" data-level="12.1" data-path="genome-environment-association.html"><a href="genome-environment-association.html#download-the-data-2"><i class="fa fa-check"></i><b>12.1</b> Download the data</a></li>
<li class="chapter" data-level="12.2" data-path="genome-environment-association.html"><a href="genome-environment-association.html#install-a-compiler-for-unixbash"><i class="fa fa-check"></i><b>12.2</b> install a compiler for Unix/bash</a></li>
<li class="chapter" data-level="12.3" data-path="genome-environment-association.html"><a href="genome-environment-association.html#install-r-packages"><i class="fa fa-check"></i><b>12.3</b> install R packages</a></li>
<li class="chapter" data-level="12.4" data-path="genome-environment-association.html"><a href="genome-environment-association.html#get-the-environmental-data"><i class="fa fa-check"></i><b>12.4</b> Get the Environmental Data</a></li>
<li class="chapter" data-level="12.5" data-path="genome-environment-association.html"><a href="genome-environment-association.html#extract-site-specific-information-from-our-envirornmental-layers"><i class="fa fa-check"></i><b>12.5</b> Extract Site Specific Information from our Envirornmental Layers</a></li>
<li class="chapter" data-level="12.6" data-path="genome-environment-association.html"><a href="genome-environment-association.html#make-a-map-of-our-environmental-data"><i class="fa fa-check"></i><b>12.6</b> make a map of our environmental data</a></li>
<li class="chapter" data-level="12.7" data-path="genome-environment-association.html"><a href="genome-environment-association.html#load-in-genetic-data"><i class="fa fa-check"></i><b>12.7</b> Load in genetic data</a></li>
<li class="chapter" data-level="12.8" data-path="genome-environment-association.html"><a href="genome-environment-association.html#exercises-2"><i class="fa fa-check"></i><b>12.8</b> Exercises</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="genome-environment-association" class="section level1" number="12">
<h1><span class="header-section-number">12</span> Genome Environment Association</h1>
<p>The lecture for this week can be found <a href="https://github.com/BayLab/MarineGenomicsSemester/blob/main/ppt/MarineGenomics_Lecture_w11.pdf">here</a>.</p>
<p>All of the code presented in this weeks class comes modified from a former Marine Genomics student and recent UC Davis graduate, Camille Rumberger! Thank you Camille!</p>
<div id="download-the-data-2" class="section level2" number="12.1">
<h2><span class="header-section-number">12.1</span> Download the data</h2>
<p>For this week we are again using data from the wonderful Xuereb et al. 2018 paper <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.14589?casa_token=aBVeQUaZ6UEAAAAA:vt9cQFbQ-65F1erD-1Uq0DAWEaZ75fDhepkPomc4RMPAqQgntIcm0btk842SxvaraM2VdlZ5nwoHUhCy">here</a>
And consistes of a vcf file with 3966 Snps from 685 pacific sea cucumbers.</p>
<div class="sourceCode" id="cb551"><pre class="sourceCode html"><code class="sourceCode html"><span id="cb551-1"><a href="genome-environment-association.html#cb551-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb551-2"><a href="genome-environment-association.html#cb551-2" aria-hidden="true" tabindex="-1"></a>wget https://raw.githubusercontent.com/BayLab/MarineGenomicsData/main/week11_semester.tar.gz</span>
<span id="cb551-3"><a href="genome-environment-association.html#cb551-3" aria-hidden="true" tabindex="-1"></a>tar -xzvf week11_semester.tar.gz</span></code></pre></div>
</div>
<div id="install-a-compiler-for-unixbash" class="section level2" number="12.2">
<h2><span class="header-section-number">12.2</span> install a compiler for Unix/bash</h2>
<p>Run this command below in the terminal. It’s necesary to install several R packages.</p>
<div class="sourceCode" id="cb552"><pre class="sourceCode html"><code class="sourceCode html"><span id="cb552-1"><a href="genome-environment-association.html#cb552-1" aria-hidden="true" tabindex="-1"></a>sudo apt install libgdal-dev</span></code></pre></div>
</div>
<div id="install-r-packages" class="section level2" number="12.3">
<h2><span class="header-section-number">12.3</span> install R packages</h2>
<div class="sourceCode" id="cb553"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb553-1"><a href="genome-environment-association.html#cb553-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb553-2"><a href="genome-environment-association.html#cb553-2" aria-hidden="true" tabindex="-1"></a><span class="co">#devtools::install_github(&quot;bcm-uga/lfmm&quot;)</span></span>
<span id="cb553-3"><a href="genome-environment-association.html#cb553-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb553-4"><a href="genome-environment-association.html#cb553-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb553-5"><a href="genome-environment-association.html#cb553-5" aria-hidden="true" tabindex="-1"></a><span class="co"># get packages</span></span>
<span id="cb553-6"><a href="genome-environment-association.html#cb553-6" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages(c(&#39;lfmm&#39;,&#39;psych&#39;,&#39;vegan&#39;,&#39;ggplot2&#39;,&#39;data.table&#39;,&#39;sdmpredictors&#39;,&#39;rgdal&#39;,&#39;raster&#39;), dependencies = T)</span></span>
<span id="cb553-7"><a href="genome-environment-association.html#cb553-7" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages(&quot;BiocManager&quot;)</span></span>
<span id="cb553-8"><a href="genome-environment-association.html#cb553-8" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install(&quot;LEA&quot;, force=T)</span></span>
<span id="cb553-9"><a href="genome-environment-association.html#cb553-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb553-10"><a href="genome-environment-association.html#cb553-10" aria-hidden="true" tabindex="-1"></a><span class="co"># call programs</span></span>
<span id="cb553-11"><a href="genome-environment-association.html#cb553-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lfmm)</span>
<span id="cb553-12"><a href="genome-environment-association.html#cb553-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;lfmm&#39; was built under R version 4.0.5</span></span>
<span id="cb553-13"><a href="genome-environment-association.html#cb553-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(psych)</span>
<span id="cb553-14"><a href="genome-environment-association.html#cb553-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;psych&#39; was built under R version 4.0.5</span></span>
<span id="cb553-15"><a href="genome-environment-association.html#cb553-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb553-16"><a href="genome-environment-association.html#cb553-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;vegan&#39; was built under R version 4.0.5</span></span>
<span id="cb553-17"><a href="genome-environment-association.html#cb553-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;permute&#39; was built under R version 4.0.5</span></span>
<span id="cb553-18"><a href="genome-environment-association.html#cb553-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;lattice&#39; was built under R version 4.0.5</span></span>
<span id="cb553-19"><a href="genome-environment-association.html#cb553-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(LEA)</span>
<span id="cb553-20"><a href="genome-environment-association.html#cb553-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb553-21"><a href="genome-environment-association.html#cb553-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;data.table&#39; was built under R version 4.0.5</span></span>
<span id="cb553-22"><a href="genome-environment-association.html#cb553-22" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sdmpredictors)</span>
<span id="cb553-23"><a href="genome-environment-association.html#cb553-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;sdmpredictors&#39; was built under R version 4.0.5</span></span>
<span id="cb553-24"><a href="genome-environment-association.html#cb553-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;raster&#39; was built under R version 4.0.5</span></span>
<span id="cb553-25"><a href="genome-environment-association.html#cb553-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;sp&#39; was built under R version 4.0.5</span></span>
<span id="cb553-26"><a href="genome-environment-association.html#cb553-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;rgdal&#39; was built under R version 4.0.5</span></span>
<span id="cb553-27"><a href="genome-environment-association.html#cb553-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(leaflet)</span>
<span id="cb553-28"><a href="genome-environment-association.html#cb553-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: package &#39;leaflet&#39; was built under R version 4.0.5</span></span>
<span id="cb553-29"><a href="genome-environment-association.html#cb553-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb553-30"><a href="genome-environment-association.html#cb553-30" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rgdal)</span>
<span id="cb553-31"><a href="genome-environment-association.html#cb553-31" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(raster)</span></code></pre></div>
</div>
<div id="get-the-environmental-data" class="section level2" number="12.4">
<h2><span class="header-section-number">12.4</span> Get the Environmental Data</h2>
<p>now we need to download the environmental data for our study. We’re just going to look at a few variables but there are many others that you could choose from here: <a href="https://www.bio-oracle.org/explore-data.php" class="uri">https://www.bio-oracle.org/explore-data.php</a></p>
<div class="sourceCode" id="cb554"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb554-1"><a href="genome-environment-association.html#cb554-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb554-2"><a href="genome-environment-association.html#cb554-2" aria-hidden="true" tabindex="-1"></a><span class="co">#load the environmental data and store it as objects that we can use later</span></span>
<span id="cb554-3"><a href="genome-environment-association.html#cb554-3" aria-hidden="true" tabindex="-1"></a>environ <span class="ot">&lt;-</span> <span class="fu">load_layers</span>(<span class="at">layercodes =</span> <span class="fu">c</span>(<span class="st">&quot;BO_chlomean&quot;</span>,<span class="st">&quot;BO_ph&quot;</span>, <span class="st">&quot;BO_salinity&quot;</span>,<span class="st">&quot;BO_sstmax&quot;</span>,<span class="st">&quot;BO_sstmean&quot;</span>))</span>
<span id="cb554-4"><a href="genome-environment-association.html#cb554-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb554-5"><a href="genome-environment-association.html#cb554-5" aria-hidden="true" tabindex="-1"></a><span class="co">#now put each layer into it&#39;s own object</span></span>
<span id="cb554-6"><a href="genome-environment-association.html#cb554-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb554-7"><a href="genome-environment-association.html#cb554-7" aria-hidden="true" tabindex="-1"></a>chlomean<span class="ot">&lt;-</span><span class="fu">load_layers</span>(<span class="st">&quot;BO_chlomean&quot;</span>)</span>
<span id="cb554-8"><a href="genome-environment-association.html#cb554-8" aria-hidden="true" tabindex="-1"></a>ph<span class="ot">&lt;-</span><span class="fu">load_layers</span>(<span class="st">&quot;BO_ph&quot;</span>)</span>
<span id="cb554-9"><a href="genome-environment-association.html#cb554-9" aria-hidden="true" tabindex="-1"></a>salinity<span class="ot">&lt;-</span><span class="fu">load_layers</span>(<span class="st">&quot;BO_salinity&quot;</span>)</span>
<span id="cb554-10"><a href="genome-environment-association.html#cb554-10" aria-hidden="true" tabindex="-1"></a>sst_max<span class="ot">&lt;-</span><span class="fu">load_layers</span>(<span class="st">&quot;BO_sstmax&quot;</span>)</span>
<span id="cb554-11"><a href="genome-environment-association.html#cb554-11" aria-hidden="true" tabindex="-1"></a>sst_mean<span class="ot">&lt;-</span><span class="fu">load_layers</span>(<span class="st">&quot;BO_sstmean&quot;</span>)</span></code></pre></div>
<p>Now we’ll read in our metadata, which has lat and lon for each sample</p>
<div class="sourceCode" id="cb555"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb555-1"><a href="genome-environment-association.html#cb555-1" aria-hidden="true" tabindex="-1"></a>meta<span class="ot">&lt;-</span><span class="fu">read.csv</span>(<span class="st">&quot;californicus_metadata.csv&quot;</span>)</span>
<span id="cb555-2"><a href="genome-environment-association.html#cb555-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb555-3"><a href="genome-environment-association.html#cb555-3" aria-hidden="true" tabindex="-1"></a><span class="co">#make a new dataframe that just contains the lat and lon and label them &quot;lat&quot; and &quot;long&quot;</span></span>
<span id="cb555-4"><a href="genome-environment-association.html#cb555-4" aria-hidden="true" tabindex="-1"></a><span class="co">#notice that the original metafile mislabels lat and lon, oops!</span></span>
<span id="cb555-5"><a href="genome-environment-association.html#cb555-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb555-6"><a href="genome-environment-association.html#cb555-6" aria-hidden="true" tabindex="-1"></a>sites<span class="ot">&lt;-</span><span class="fu">cbind</span>(<span class="st">&quot;lat&quot;</span> <span class="ot">=</span>  meta<span class="sc">$</span>LONG, <span class="st">&quot;lon&quot;</span> <span class="ot">=</span> meta<span class="sc">$</span>LAT)</span>
<span id="cb555-7"><a href="genome-environment-association.html#cb555-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb555-8"><a href="genome-environment-association.html#cb555-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sites)</span>
<span id="cb555-9"><a href="genome-environment-association.html#cb555-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;            lat     lon</span></span>
<span id="cb555-10"><a href="genome-environment-association.html#cb555-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1,] -126.5646 50.5212</span></span>
<span id="cb555-11"><a href="genome-environment-association.html#cb555-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2,] -126.5646 50.5212</span></span>
<span id="cb555-12"><a href="genome-environment-association.html#cb555-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [3,] -126.5646 50.5212</span></span>
<span id="cb555-13"><a href="genome-environment-association.html#cb555-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [4,] -126.5646 50.5212</span></span>
<span id="cb555-14"><a href="genome-environment-association.html#cb555-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [5,] -126.5646 50.5212</span></span>
<span id="cb555-15"><a href="genome-environment-association.html#cb555-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [6,] -126.5646 50.5212</span></span></code></pre></div>
</div>
<div id="extract-site-specific-information-from-our-envirornmental-layers" class="section level2" number="12.5">
<h2><span class="header-section-number">12.5</span> Extract Site Specific Information from our Envirornmental Layers</h2>
<div class="sourceCode" id="cb556"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb556-1"><a href="genome-environment-association.html#cb556-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb556-2"><a href="genome-environment-association.html#cb556-2" aria-hidden="true" tabindex="-1"></a>sites_environ <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Name=</span>sites, <span class="fu">extract</span>(environ,sites))</span>
<span id="cb556-3"><a href="genome-environment-association.html#cb556-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sites_environ)</span>
<span id="cb556-4"><a href="genome-environment-association.html#cb556-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Name.lat Name.lon BO_chlomean BO_ph BO_salinity BO_sstmax BO_sstmean</span></span>
<span id="cb556-5"><a href="genome-environment-association.html#cb556-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 -126.5646  50.5212        3.85 8.212      30.496    12.137        8.1</span></span>
<span id="cb556-6"><a href="genome-environment-association.html#cb556-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 -126.5646  50.5212        3.85 8.212      30.496    12.137        8.1</span></span>
<span id="cb556-7"><a href="genome-environment-association.html#cb556-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 -126.5646  50.5212        3.85 8.212      30.496    12.137        8.1</span></span>
<span id="cb556-8"><a href="genome-environment-association.html#cb556-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4 -126.5646  50.5212        3.85 8.212      30.496    12.137        8.1</span></span>
<span id="cb556-9"><a href="genome-environment-association.html#cb556-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5 -126.5646  50.5212        3.85 8.212      30.496    12.137        8.1</span></span>
<span id="cb556-10"><a href="genome-environment-association.html#cb556-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6 -126.5646  50.5212        3.85 8.212      30.496    12.137        8.1</span></span></code></pre></div>
<p>That produces for us a site specific environment file. We now need to convert this file into a format that we can save it as a matrix and export it as an environment file (i.e., in the format that the next step need it to be in)</p>
<div class="sourceCode" id="cb557"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb557-1"><a href="genome-environment-association.html#cb557-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb557-2"><a href="genome-environment-association.html#cb557-2" aria-hidden="true" tabindex="-1"></a><span class="co">#remove lat and lon and convert to matrix</span></span>
<span id="cb557-3"><a href="genome-environment-association.html#cb557-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb557-4"><a href="genome-environment-association.html#cb557-4" aria-hidden="true" tabindex="-1"></a>sites_environ_matrix<span class="ot">&lt;-</span><span class="fu">as.matrix</span>(sites_environ[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)])</span>
<span id="cb557-5"><a href="genome-environment-association.html#cb557-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb557-6"><a href="genome-environment-association.html#cb557-6" aria-hidden="true" tabindex="-1"></a><span class="co">#remove any Na&#39;s</span></span>
<span id="cb557-7"><a href="genome-environment-association.html#cb557-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb557-8"><a href="genome-environment-association.html#cb557-8" aria-hidden="true" tabindex="-1"></a>sites_environ_matrix_nas<span class="ot">&lt;-</span><span class="fu">na.omit</span>(sites_environ_matrix)</span>
<span id="cb557-9"><a href="genome-environment-association.html#cb557-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb557-10"><a href="genome-environment-association.html#cb557-10" aria-hidden="true" tabindex="-1"></a><span class="co">#write the file to an env file</span></span>
<span id="cb557-11"><a href="genome-environment-association.html#cb557-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb557-12"><a href="genome-environment-association.html#cb557-12" aria-hidden="true" tabindex="-1"></a><span class="fu">write.env</span>(sites_environ_matrix_nas, <span class="st">&quot;sites_environ_matrix.env&quot;</span>)</span>
<span id="cb557-13"><a href="genome-environment-association.html#cb557-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;sites_environ_matrix.env&quot;</span></span></code></pre></div>
</div>
<div id="make-a-map-of-our-environmental-data" class="section level2" number="12.6">
<h2><span class="header-section-number">12.6</span> make a map of our environmental data</h2>
<div class="sourceCode" id="cb558"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb558-1"><a href="genome-environment-association.html#cb558-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sdmpredictors)</span>
<span id="cb558-2"><a href="genome-environment-association.html#cb558-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb558-3"><a href="genome-environment-association.html#cb558-3" aria-hidden="true" tabindex="-1"></a><span class="co">#download a raster file for env variable that we want to look at</span></span>
<span id="cb558-4"><a href="genome-environment-association.html#cb558-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb558-5"><a href="genome-environment-association.html#cb558-5" aria-hidden="true" tabindex="-1"></a><span class="co">#what are are min amd max lat and lon</span></span>
<span id="cb558-6"><a href="genome-environment-association.html#cb558-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb558-7"><a href="genome-environment-association.html#cb558-7" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(sites_environ<span class="sc">$</span>Name.lon)</span>
<span id="cb558-8"><a href="genome-environment-association.html#cb558-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 48.40822 57.82796</span></span>
<span id="cb558-9"><a href="genome-environment-association.html#cb558-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb558-10"><a href="genome-environment-association.html#cb558-10" aria-hidden="true" tabindex="-1"></a><span class="co">#crop file to fit the area we want</span></span>
<span id="cb558-11"><a href="genome-environment-association.html#cb558-11" aria-hidden="true" tabindex="-1"></a>ne.pacific<span class="ot">&lt;-</span><span class="fu">extent</span>(<span class="sc">-</span><span class="dv">140</span>, <span class="sc">-</span><span class="dv">120</span>, <span class="dv">40</span>, <span class="dv">60</span>)</span>
<span id="cb558-12"><a href="genome-environment-association.html#cb558-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb558-13"><a href="genome-environment-association.html#cb558-13" aria-hidden="true" tabindex="-1"></a>sst_max.crop<span class="ot">&lt;-</span><span class="fu">crop</span>(sst_max, ne.pacific)</span>
<span id="cb558-14"><a href="genome-environment-association.html#cb558-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb558-15"><a href="genome-environment-association.html#cb558-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb558-16"><a href="genome-environment-association.html#cb558-16" aria-hidden="true" tabindex="-1"></a><span class="co">#make a nice color ramp and plot the map</span></span>
<span id="cb558-17"><a href="genome-environment-association.html#cb558-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb558-18"><a href="genome-environment-association.html#cb558-18" aria-hidden="true" tabindex="-1"></a>my.colors <span class="ot">=</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;#5E85B8&quot;</span>,<span class="st">&quot;#EDF0C0&quot;</span>,<span class="st">&quot;#C13127&quot;</span>))</span>
<span id="cb558-19"><a href="genome-environment-association.html#cb558-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sst_max.crop,<span class="at">col=</span><span class="fu">my.colors</span>(<span class="dv">1000</span>),<span class="at">axes=</span><span class="cn">FALSE</span>, <span class="at">box=</span><span class="cn">FALSE</span>)</span>
<span id="cb558-20"><a href="genome-environment-association.html#cb558-20" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">cex.sub =</span> <span class="fl">1.25</span>, <span class="at">sub =</span> <span class="st">&quot;Maximum temperature at the sea surface (ºC)&quot;</span>)</span></code></pre></div>
<p><img src="Semester_files/figure-html/geamap-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="load-in-genetic-data" class="section level2" number="12.7">
<h2><span class="header-section-number">12.7</span> Load in genetic data</h2>
<p>Now we need to load in our genetic data. We have data in a vcf file format and the first few steps will be to convert it to the format needed by the lfmm package.</p>
<div class="sourceCode" id="cb559"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb559-1"><a href="genome-environment-association.html#cb559-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb559-2"><a href="genome-environment-association.html#cb559-2" aria-hidden="true" tabindex="-1"></a><span class="co">#convert our vcf file to lfmm</span></span>
<span id="cb559-3"><a href="genome-environment-association.html#cb559-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb559-4"><a href="genome-environment-association.html#cb559-4" aria-hidden="true" tabindex="-1"></a><span class="fu">vcf2lfmm</span>(<span class="st">&#39;filtered_3699snps_californicus_685inds.recode.vcf&#39;</span>, <span class="st">&#39;filtered_3699snps_californicus.lfmm&#39;</span>)</span>
<span id="cb559-5"><a href="genome-environment-association.html#cb559-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb559-6"><a href="genome-environment-association.html#cb559-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  - number of detected individuals:   685</span></span>
<span id="cb559-7"><a href="genome-environment-association.html#cb559-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  - number of detected loci:      3699</span></span>
<span id="cb559-8"><a href="genome-environment-association.html#cb559-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb559-9"><a href="genome-environment-association.html#cb559-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; For SNP info, please check ./filtered_3699snps_californicus_685inds.recode.vcfsnp.</span></span>
<span id="cb559-10"><a href="genome-environment-association.html#cb559-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb559-11"><a href="genome-environment-association.html#cb559-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0 line(s) were removed because these are not SNPs.</span></span>
<span id="cb559-12"><a href="genome-environment-association.html#cb559-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Please, check ./filtered_3699snps_californicus_685inds.recode.removed file, for more informations.</span></span>
<span id="cb559-13"><a href="genome-environment-association.html#cb559-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb559-14"><a href="genome-environment-association.html#cb559-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb559-15"><a href="genome-environment-association.html#cb559-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  - number of detected individuals:   685</span></span>
<span id="cb559-16"><a href="genome-environment-association.html#cb559-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  - number of detected loci:      3699</span></span>
<span id="cb559-17"><a href="genome-environment-association.html#cb559-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;./filtered_3699snps_californicus_685inds.recode.lfmm&quot;</span></span>
<span id="cb559-18"><a href="genome-environment-association.html#cb559-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb559-19"><a href="genome-environment-association.html#cb559-19" aria-hidden="true" tabindex="-1"></a>sea_cuc_lfmm<span class="ot">&lt;-</span><span class="fu">read.lfmm</span>(<span class="st">&#39;filtered_3699snps_californicus_685inds.recode.lfmm&#39;</span>)</span>
<span id="cb559-20"><a href="genome-environment-association.html#cb559-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb559-21"><a href="genome-environment-association.html#cb559-21" aria-hidden="true" tabindex="-1"></a><span class="co">#and convert to geno</span></span>
<span id="cb559-22"><a href="genome-environment-association.html#cb559-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb559-23"><a href="genome-environment-association.html#cb559-23" aria-hidden="true" tabindex="-1"></a><span class="fu">lfmm2geno</span>(<span class="st">&#39;filtered_3699snps_californicus_685inds.recode.lfmm&#39;</span>, <span class="st">&#39;filtered_3699snps_californicus_685inds.recode.geno&#39;</span>)</span>
<span id="cb559-24"><a href="genome-environment-association.html#cb559-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb559-25"><a href="genome-environment-association.html#cb559-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  - number of detected individuals:   685</span></span>
<span id="cb559-26"><a href="genome-environment-association.html#cb559-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  - number of detected loci:      3699</span></span>
<span id="cb559-27"><a href="genome-environment-association.html#cb559-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;filtered_3699snps_californicus_685inds.recode.geno&quot;</span></span>
<span id="cb559-28"><a href="genome-environment-association.html#cb559-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb559-29"><a href="genome-environment-association.html#cb559-29" aria-hidden="true" tabindex="-1"></a><span class="co">#read in geno file</span></span>
<span id="cb559-30"><a href="genome-environment-association.html#cb559-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb559-31"><a href="genome-environment-association.html#cb559-31" aria-hidden="true" tabindex="-1"></a>sea_cuc_geno<span class="ot">&lt;-</span><span class="fu">read.geno</span>(<span class="st">&#39;filtered_3699snps_californicus_685inds.recode.geno&#39;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb560"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb560-1"><a href="genome-environment-association.html#cb560-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb560-2"><a href="genome-environment-association.html#cb560-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb560-3"><a href="genome-environment-association.html#cb560-3" aria-hidden="true" tabindex="-1"></a><span class="co"># create a snmf object</span></span>
<span id="cb560-4"><a href="genome-environment-association.html#cb560-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb560-5"><a href="genome-environment-association.html#cb560-5" aria-hidden="true" tabindex="-1"></a>sea_cuc200.snmf <span class="ot">&lt;-</span> <span class="fu">snmf</span>(<span class="st">&quot;filtered_3699snps_californicus_685.geno&quot;</span>, <span class="at">K =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">entropy=</span>T, <span class="at">ploidy =</span> <span class="dv">2</span>, <span class="at">project =</span> <span class="st">&quot;new&quot;</span>)</span>
<span id="cb560-6"><a href="genome-environment-association.html#cb560-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning in normalizePath(path.expand(path), winslash, mustWork): path[1]=&quot;filtered_3699snps_californicus_685.geno&quot;: The</span></span>
<span id="cb560-7"><a href="genome-environment-association.html#cb560-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; system cannot find the file specified</span></span>
<span id="cb560-8"><a href="genome-environment-association.html#cb560-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning in normalizePath(path.expand(path), winslash, mustWork): path[1]=&quot;C:</span></span>
<span id="cb560-9"><a href="genome-environment-association.html#cb560-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; \Users\SAPCaps\MarineGenomicsSemester\data\Week11_gea\filtered_3699snps_californicus_685.geno&quot;: The system cannot find</span></span>
<span id="cb560-10"><a href="genome-environment-association.html#cb560-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; the file specified</span></span>
<span id="cb560-11"><a href="genome-environment-association.html#cb560-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb560-12"><a href="genome-environment-association.html#cb560-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning in normalizePath(path.expand(path), winslash, mustWork): path[1]=&quot;C:</span></span>
<span id="cb560-13"><a href="genome-environment-association.html#cb560-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; \Users\SAPCaps\MarineGenomicsSemester\data\Week11_gea\filtered_3699snps_californicus_685.geno&quot;: The system cannot find</span></span>
<span id="cb560-14"><a href="genome-environment-association.html#cb560-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; the file specified</span></span>
<span id="cb560-15"><a href="genome-environment-association.html#cb560-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The project is saved into :</span></span>
<span id="cb560-16"><a href="genome-environment-association.html#cb560-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  filtered_3699snps_californicus_685.snmfProject </span></span>
<span id="cb560-17"><a href="genome-environment-association.html#cb560-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb560-18"><a href="genome-environment-association.html#cb560-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; To load the project, use:</span></span>
<span id="cb560-19"><a href="genome-environment-association.html#cb560-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  project = load.snmfProject(&quot;filtered_3699snps_californicus_685.snmfProject&quot;)</span></span>
<span id="cb560-20"><a href="genome-environment-association.html#cb560-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb560-21"><a href="genome-environment-association.html#cb560-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; To remove the project, use:</span></span>
<span id="cb560-22"><a href="genome-environment-association.html#cb560-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  remove.snmfProject(&quot;filtered_3699snps_californicus_685.snmfProject&quot;)</span></span>
<span id="cb560-23"><a href="genome-environment-association.html#cb560-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb560-24"><a href="genome-environment-association.html#cb560-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 1119183035</span></span>
<span id="cb560-25"><a href="genome-environment-association.html#cb560-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;*************************************&quot;</span></span>
<span id="cb560-26"><a href="genome-environment-association.html#cb560-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;*          create.dataset            *&quot;</span></span>
<span id="cb560-27"><a href="genome-environment-association.html#cb560-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;*************************************&quot;</span></span>
<span id="cb560-28"><a href="genome-environment-association.html#cb560-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb560-29"><a href="genome-environment-association.html#cb560-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ERROR: unable to open file C:\Users\SAPCaps\MarineGenomicsSemester\data\Week11_gea\filtered_3699snps_californicus_685.geno. Please, check that the name of the file you provided is correct.</span></span>
<span id="cb560-30"><a href="genome-environment-association.html#cb560-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Error in create.dataset(input.file, masked.file, s, percentage): internal error in trio library</span></span>
<span id="cb560-31"><a href="genome-environment-association.html#cb560-31" aria-hidden="true" tabindex="-1"></a>project<span class="ot">=</span><span class="fu">load.snmfProject</span>(<span class="st">&quot;filtered_3699snps_californicus.snmfProject&quot;</span>)</span>
<span id="cb560-32"><a href="genome-environment-association.html#cb560-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb560-33"><a href="genome-environment-association.html#cb560-33" aria-hidden="true" tabindex="-1"></a><span class="co">#plot values of cross-entropy criteron of k</span></span>
<span id="cb560-34"><a href="genome-environment-association.html#cb560-34" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sea_cuc200.snmf)</span>
<span id="cb560-35"><a href="genome-environment-association.html#cb560-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Error in h(simpleError(msg, call)): error in evaluating the argument &#39;x&#39; in selecting a method for function &#39;plot&#39;: object &#39;sea_cuc200.snmf&#39; not found</span></span></code></pre></div>
<p>This plot tells us which value of K we should use when running the next analyses. We’re looking for a region where the cross-entropy is lowest and ideally it plateaus. We don’t get a plateau in our case but you can see our lowest value is at 3. So we would use a K of 3 in the next step.</p>
<p>However, for the purposes of time we’re going to do the next step with a k=1 and a rep=1.</p>
<pre><code>#&gt; The project is saved into :
#&gt;  filtered_3699snps_californicus_685inds.lfmmProject 
#&gt; 
#&gt; To load the project, use:
#&gt;  project = load.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; To remove the project, use:
#&gt;  remove.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; [1] &quot;********************************&quot;
#&gt; [1] &quot;* K = 1  repetition 1  d = 1   *&quot;
#&gt; [1] &quot;********************************&quot;
#&gt; Summary of the options:
#&gt; 
#&gt;         -n (number of individuals)      685
#&gt;         -L (number of loci)             3699
#&gt;         -K (number of latent factors)   1
#&gt;         -o (output file)                filtered_3699snps_californicus_685inds.lfmm/K1/run1/filtered_3699snps_californicus_685inds.recode_r1
#&gt;         -i (number of iterations)       10000
#&gt;         -b (burnin)                     5000
#&gt;         -s (seed random init)           610934770
#&gt;         -p (number of processes (CPU))  1
#&gt;         -x (genotype file)              filtered_3699snps_californicus_685inds.recode.lfmm
#&gt;         -v (variable file)              sites_environ_matrix.env
#&gt;         -D (number of covariables)      5
#&gt;         -d (the dth covariable)         1
#&gt; 
#&gt; Read variable file:
#&gt;      sites_environ_matrix.env        OK.
#&gt; 
#&gt; Read genotype file:
#&gt;      filtered_3699snps_californicus_685inds.recode.lfmm      OK.
#&gt; 
#&gt; &lt;&lt;&lt;&lt;
#&gt;   Analyse for variable 1
#&gt; 
#&gt;      Start of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  [                                                                           ]
#&gt;  [===========================================================================]
#&gt; 
#&gt;      End of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  ED:2533818.132   DIC: 2533819.189 
#&gt; 
#&gt;  The statistics for the run are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run1/filtered_3699snps_californicus_685inds.recode_r1_s1.1.dic.
#&gt; 
#&gt;  The zscores for variable 1 are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run1/filtered_3699snps_californicus_685inds.recode_r1_s1.1.zscore.
#&gt;  The columns are: zscores, -log10(p-values), p-values.
#&gt; 
#&gt;  -------------------------
#&gt;  The execution for variable 1 worked without error.
#&gt; &gt;&gt;&gt;&gt;
#&gt; 
#&gt; The project is saved into :
#&gt;  filtered_3699snps_californicus_685inds.lfmmProject 
#&gt; 
#&gt; To load the project, use:
#&gt;  project = load.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; To remove the project, use:
#&gt;  remove.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; [1] &quot;********************************&quot;
#&gt; [1] &quot;* K = 1  repetition 1  d = 2   *&quot;
#&gt; [1] &quot;********************************&quot;
#&gt; Summary of the options:
#&gt; 
#&gt;         -n (number of individuals)      685
#&gt;         -L (number of loci)             3699
#&gt;         -K (number of latent factors)   1
#&gt;         -o (output file)                filtered_3699snps_californicus_685inds.lfmm/K1/run2/filtered_3699snps_californicus_685inds.recode_r2
#&gt;         -i (number of iterations)       10000
#&gt;         -b (burnin)                     5000
#&gt;         -s (seed random init)           610934770
#&gt;         -p (number of processes (CPU))  1
#&gt;         -x (genotype file)              filtered_3699snps_californicus_685inds.recode.lfmm
#&gt;         -v (variable file)              sites_environ_matrix.env
#&gt;         -D (number of covariables)      5
#&gt;         -d (the dth covariable)         2
#&gt; 
#&gt; Read variable file:
#&gt;      sites_environ_matrix.env        OK.
#&gt; 
#&gt; Read genotype file:
#&gt;      filtered_3699snps_californicus_685inds.recode.lfmm      OK.
#&gt; 
#&gt; &lt;&lt;&lt;&lt;
#&gt;   Analyse for variable 2
#&gt; 
#&gt;      Start of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  [                                                                           ]
#&gt;  [===========================================================================]
#&gt; 
#&gt;      End of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  ED:2533817.996   DIC: 2533818.941 
#&gt; 
#&gt;  The statistics for the run are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run2/filtered_3699snps_californicus_685inds.recode_r2_s2.1.dic.
#&gt; 
#&gt;  The zscores for variable 2 are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run2/filtered_3699snps_californicus_685inds.recode_r2_s2.1.zscore.
#&gt;  The columns are: zscores, -log10(p-values), p-values.
#&gt; 
#&gt;  -------------------------
#&gt;  The execution for variable 2 worked without error.
#&gt; &gt;&gt;&gt;&gt;
#&gt; 
#&gt; The project is saved into :
#&gt;  filtered_3699snps_californicus_685inds.lfmmProject 
#&gt; 
#&gt; To load the project, use:
#&gt;  project = load.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; To remove the project, use:
#&gt;  remove.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; [1] &quot;********************************&quot;
#&gt; [1] &quot;* K = 1  repetition 1  d = 3   *&quot;
#&gt; [1] &quot;********************************&quot;
#&gt; Summary of the options:
#&gt; 
#&gt;         -n (number of individuals)      685
#&gt;         -L (number of loci)             3699
#&gt;         -K (number of latent factors)   1
#&gt;         -o (output file)                filtered_3699snps_californicus_685inds.lfmm/K1/run3/filtered_3699snps_californicus_685inds.recode_r3
#&gt;         -i (number of iterations)       10000
#&gt;         -b (burnin)                     5000
#&gt;         -s (seed random init)           610934770
#&gt;         -p (number of processes (CPU))  1
#&gt;         -x (genotype file)              filtered_3699snps_californicus_685inds.recode.lfmm
#&gt;         -v (variable file)              sites_environ_matrix.env
#&gt;         -D (number of covariables)      5
#&gt;         -d (the dth covariable)         3
#&gt; 
#&gt; Read variable file:
#&gt;      sites_environ_matrix.env        OK.
#&gt; 
#&gt; Read genotype file:
#&gt;      filtered_3699snps_californicus_685inds.recode.lfmm      OK.
#&gt; 
#&gt; &lt;&lt;&lt;&lt;
#&gt;   Analyse for variable 3
#&gt; 
#&gt;      Start of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  [                                                                           ]
#&gt;  [===========================================================================]
#&gt; 
#&gt;      End of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  ED:2533818.069   DIC: 2533819.06 
#&gt; 
#&gt;  The statistics for the run are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run3/filtered_3699snps_californicus_685inds.recode_r3_s3.1.dic.
#&gt; 
#&gt;  The zscores for variable 3 are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run3/filtered_3699snps_californicus_685inds.recode_r3_s3.1.zscore.
#&gt;  The columns are: zscores, -log10(p-values), p-values.
#&gt; 
#&gt;  -------------------------
#&gt;  The execution for variable 3 worked without error.
#&gt; &gt;&gt;&gt;&gt;
#&gt; 
#&gt; The project is saved into :
#&gt;  filtered_3699snps_californicus_685inds.lfmmProject 
#&gt; 
#&gt; To load the project, use:
#&gt;  project = load.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; To remove the project, use:
#&gt;  remove.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; [1] &quot;********************************&quot;
#&gt; [1] &quot;* K = 1  repetition 1  d = 4   *&quot;
#&gt; [1] &quot;********************************&quot;
#&gt; Summary of the options:
#&gt; 
#&gt;         -n (number of individuals)      685
#&gt;         -L (number of loci)             3699
#&gt;         -K (number of latent factors)   1
#&gt;         -o (output file)                filtered_3699snps_californicus_685inds.lfmm/K1/run4/filtered_3699snps_californicus_685inds.recode_r4
#&gt;         -i (number of iterations)       10000
#&gt;         -b (burnin)                     5000
#&gt;         -s (seed random init)           610934770
#&gt;         -p (number of processes (CPU))  1
#&gt;         -x (genotype file)              filtered_3699snps_californicus_685inds.recode.lfmm
#&gt;         -v (variable file)              sites_environ_matrix.env
#&gt;         -D (number of covariables)      5
#&gt;         -d (the dth covariable)         4
#&gt; 
#&gt; Read variable file:
#&gt;      sites_environ_matrix.env        OK.
#&gt; 
#&gt; Read genotype file:
#&gt;      filtered_3699snps_californicus_685inds.recode.lfmm      OK.
#&gt; 
#&gt; &lt;&lt;&lt;&lt;
#&gt;   Analyse for variable 4
#&gt; 
#&gt;      Start of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  [                                                                           ]
#&gt;  [===========================================================================]
#&gt; 
#&gt;      End of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  ED:2533818.155   DIC: 2533819.217 
#&gt; 
#&gt;  The statistics for the run are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run4/filtered_3699snps_californicus_685inds.recode_r4_s4.1.dic.
#&gt; 
#&gt;  The zscores for variable 4 are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run4/filtered_3699snps_californicus_685inds.recode_r4_s4.1.zscore.
#&gt;  The columns are: zscores, -log10(p-values), p-values.
#&gt; 
#&gt;  -------------------------
#&gt;  The execution for variable 4 worked without error.
#&gt; &gt;&gt;&gt;&gt;
#&gt; 
#&gt; The project is saved into :
#&gt;  filtered_3699snps_californicus_685inds.lfmmProject 
#&gt; 
#&gt; To load the project, use:
#&gt;  project = load.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; To remove the project, use:
#&gt;  remove.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; [1] &quot;********************************&quot;
#&gt; [1] &quot;* K = 1  repetition 1  d = 5   *&quot;
#&gt; [1] &quot;********************************&quot;
#&gt; Summary of the options:
#&gt; 
#&gt;         -n (number of individuals)      685
#&gt;         -L (number of loci)             3699
#&gt;         -K (number of latent factors)   1
#&gt;         -o (output file)                filtered_3699snps_californicus_685inds.lfmm/K1/run5/filtered_3699snps_californicus_685inds.recode_r5
#&gt;         -i (number of iterations)       10000
#&gt;         -b (burnin)                     5000
#&gt;         -s (seed random init)           610934770
#&gt;         -p (number of processes (CPU))  1
#&gt;         -x (genotype file)              filtered_3699snps_californicus_685inds.recode.lfmm
#&gt;         -v (variable file)              sites_environ_matrix.env
#&gt;         -D (number of covariables)      5
#&gt;         -d (the dth covariable)         5
#&gt; 
#&gt; Read variable file:
#&gt;      sites_environ_matrix.env        OK.
#&gt; 
#&gt; Read genotype file:
#&gt;      filtered_3699snps_californicus_685inds.recode.lfmm      OK.
#&gt; 
#&gt; &lt;&lt;&lt;&lt;
#&gt;   Analyse for variable 5
#&gt; 
#&gt;      Start of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  [                                                                           ]
#&gt;  [===========================================================================]
#&gt; 
#&gt;      End of the Gibbs Sampler algorithm.
#&gt; 
#&gt;  ED:2533818.091   DIC: 2533819.071 
#&gt; 
#&gt;  The statistics for the run are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run5/filtered_3699snps_californicus_685inds.recode_r5_s5.1.dic.
#&gt; 
#&gt;  The zscores for variable 5 are registered in:
#&gt;          filtered_3699snps_californicus_685inds.lfmm/K1/run5/filtered_3699snps_californicus_685inds.recode_r5_s5.1.zscore.
#&gt;  The columns are: zscores, -log10(p-values), p-values.
#&gt; 
#&gt;  -------------------------
#&gt;  The execution for variable 5 worked without error.
#&gt; &gt;&gt;&gt;&gt;
#&gt; 
#&gt; The project is saved into :
#&gt;  filtered_3699snps_californicus_685inds.lfmmProject 
#&gt; 
#&gt; To load the project, use:
#&gt;  project = load.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)
#&gt; 
#&gt; To remove the project, use:
#&gt;  remove.lfmmProject(&quot;filtered_3699snps_californicus_685inds.lfmmProject&quot;)</code></pre>
<p>Now we can check and see which of our environmental variables is association with genomic variation</p>
<div class="sourceCode" id="cb562"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb562-1"><a href="genome-environment-association.html#cb562-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get the z scores from each run</span></span>
<span id="cb562-2"><a href="genome-environment-association.html#cb562-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb562-3"><a href="genome-environment-association.html#cb562-3" aria-hidden="true" tabindex="-1"></a>zs1 <span class="ot">&lt;-</span> <span class="fu">z.scores</span>(sea_cuc_runlfmm, <span class="at">K=</span><span class="dv">1</span>, <span class="at">d=</span><span class="dv">1</span>)</span>
<span id="cb562-4"><a href="genome-environment-association.html#cb562-4" aria-hidden="true" tabindex="-1"></a>zs2 <span class="ot">&lt;-</span> <span class="fu">z.scores</span>(sea_cuc_runlfmm, <span class="at">K=</span><span class="dv">1</span>, <span class="at">d=</span><span class="dv">2</span>)</span>
<span id="cb562-5"><a href="genome-environment-association.html#cb562-5" aria-hidden="true" tabindex="-1"></a>zs3 <span class="ot">&lt;-</span> <span class="fu">z.scores</span>(sea_cuc_runlfmm, <span class="at">K=</span><span class="dv">1</span>, <span class="at">d=</span><span class="dv">3</span>)</span>
<span id="cb562-6"><a href="genome-environment-association.html#cb562-6" aria-hidden="true" tabindex="-1"></a>zs4 <span class="ot">&lt;-</span> <span class="fu">z.scores</span>(sea_cuc_runlfmm, <span class="at">K=</span><span class="dv">1</span>, <span class="at">d=</span><span class="dv">4</span>)</span>
<span id="cb562-7"><a href="genome-environment-association.html#cb562-7" aria-hidden="true" tabindex="-1"></a>zs5 <span class="ot">&lt;-</span> <span class="fu">z.scores</span>(sea_cuc_runlfmm, <span class="at">K=</span><span class="dv">1</span>, <span class="at">d=</span><span class="dv">5</span>)</span>
<span id="cb562-8"><a href="genome-environment-association.html#cb562-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb562-9"><a href="genome-environment-association.html#cb562-9" aria-hidden="true" tabindex="-1"></a><span class="co"># compute the genomic inflation factor lambda</span></span>
<span id="cb562-10"><a href="genome-environment-association.html#cb562-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb562-11"><a href="genome-environment-association.html#cb562-11" aria-hidden="true" tabindex="-1"></a>lambda1<span class="ot">=</span><span class="fu">median</span>(zs1<span class="sc">^</span><span class="dv">2</span>)<span class="sc">/</span><span class="fu">qchisq</span>(<span class="fl">0.5</span>, <span class="at">df=</span><span class="dv">1</span>)</span>
<span id="cb562-12"><a href="genome-environment-association.html#cb562-12" aria-hidden="true" tabindex="-1"></a>lambda2<span class="ot">=</span><span class="fu">median</span>(zs2<span class="sc">^</span><span class="dv">2</span>)<span class="sc">/</span><span class="fu">qchisq</span>(<span class="fl">0.5</span>, <span class="at">df=</span><span class="dv">1</span>)</span>
<span id="cb562-13"><a href="genome-environment-association.html#cb562-13" aria-hidden="true" tabindex="-1"></a>lambda3<span class="ot">=</span><span class="fu">median</span>(zs3<span class="sc">^</span><span class="dv">2</span>)<span class="sc">/</span><span class="fu">qchisq</span>(<span class="fl">0.5</span>, <span class="at">df=</span><span class="dv">1</span>)</span>
<span id="cb562-14"><a href="genome-environment-association.html#cb562-14" aria-hidden="true" tabindex="-1"></a>lambda4<span class="ot">=</span><span class="fu">median</span>(zs4<span class="sc">^</span><span class="dv">2</span>)<span class="sc">/</span><span class="fu">qchisq</span>(<span class="fl">0.5</span>, <span class="at">df=</span><span class="dv">1</span>)</span>
<span id="cb562-15"><a href="genome-environment-association.html#cb562-15" aria-hidden="true" tabindex="-1"></a>lambda5<span class="ot">=</span><span class="fu">median</span>(zs5<span class="sc">^</span><span class="dv">2</span>)<span class="sc">/</span><span class="fu">qchisq</span>(<span class="fl">0.5</span>, <span class="at">df=</span><span class="dv">1</span>)</span>
<span id="cb562-16"><a href="genome-environment-association.html#cb562-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb562-17"><a href="genome-environment-association.html#cb562-17" aria-hidden="true" tabindex="-1"></a><span class="co">#then compute adjusted p-values</span></span>
<span id="cb562-18"><a href="genome-environment-association.html#cb562-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb562-19"><a href="genome-environment-association.html#cb562-19" aria-hidden="true" tabindex="-1"></a>adj.p.val1<span class="ot">&lt;-</span><span class="fu">pchisq</span>(zs1<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span>lambda1, <span class="at">df=</span><span class="dv">1</span>, <span class="at">lower=</span>F)</span>
<span id="cb562-20"><a href="genome-environment-association.html#cb562-20" aria-hidden="true" tabindex="-1"></a>adj.p.val2<span class="ot">&lt;-</span><span class="fu">pchisq</span>(zs2<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span>lambda2, <span class="at">df=</span><span class="dv">1</span>, <span class="at">lower=</span>F)</span>
<span id="cb562-21"><a href="genome-environment-association.html#cb562-21" aria-hidden="true" tabindex="-1"></a>adj.p.val3<span class="ot">&lt;-</span><span class="fu">pchisq</span>(zs3<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span>lambda3, <span class="at">df=</span><span class="dv">1</span>, <span class="at">lower=</span>F)</span>
<span id="cb562-22"><a href="genome-environment-association.html#cb562-22" aria-hidden="true" tabindex="-1"></a>adj.p.val4<span class="ot">&lt;-</span><span class="fu">pchisq</span>(zs4<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span>lambda4, <span class="at">df=</span><span class="dv">1</span>, <span class="at">lower=</span>F)</span>
<span id="cb562-23"><a href="genome-environment-association.html#cb562-23" aria-hidden="true" tabindex="-1"></a>adj.p.val5<span class="ot">&lt;-</span><span class="fu">pchisq</span>(zs5<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span>lambda5, <span class="at">df=</span><span class="dv">1</span>, <span class="at">lower=</span>F)</span>
<span id="cb562-24"><a href="genome-environment-association.html#cb562-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb562-25"><a href="genome-environment-association.html#cb562-25" aria-hidden="true" tabindex="-1"></a><span class="co">#then plot that p-value</span></span>
<span id="cb562-26"><a href="genome-environment-association.html#cb562-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb562-27"><a href="genome-environment-association.html#cb562-27" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(adj.p.val1, <span class="at">col=</span><span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<p><img src="Semester_files/figure-html/unnamed-chunk-113-1.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb563"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb563-1"><a href="genome-environment-association.html#cb563-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(adj.p.val2, <span class="at">col=</span><span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<p><img src="Semester_files/figure-html/unnamed-chunk-113-2.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb564"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb564-1"><a href="genome-environment-association.html#cb564-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(adj.p.val3, <span class="at">col=</span><span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<p><img src="Semester_files/figure-html/unnamed-chunk-113-3.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb565"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb565-1"><a href="genome-environment-association.html#cb565-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(adj.p.val4, <span class="at">col=</span><span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<p><img src="Semester_files/figure-html/unnamed-chunk-113-4.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb566"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb566-1"><a href="genome-environment-association.html#cb566-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(adj.p.val5, <span class="at">col=</span><span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<p><img src="Semester_files/figure-html/unnamed-chunk-113-5.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb567"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb567-1"><a href="genome-environment-association.html#cb567-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb567-2"><a href="genome-environment-association.html#cb567-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb567-3"><a href="genome-environment-association.html#cb567-3" aria-hidden="true" tabindex="-1"></a><span class="co"># control of false discoveries</span></span>
<span id="cb567-4"><a href="genome-environment-association.html#cb567-4" aria-hidden="true" tabindex="-1"></a><span class="co"># to correct for multiple testing, we can apply the Benjamini-Hochberg algorithm</span></span>
<span id="cb567-5"><a href="genome-environment-association.html#cb567-5" aria-hidden="true" tabindex="-1"></a><span class="co"># L is number of loci</span></span>
<span id="cb567-6"><a href="genome-environment-association.html#cb567-6" aria-hidden="true" tabindex="-1"></a>L<span class="ot">=</span><span class="dv">3699</span></span>
<span id="cb567-7"><a href="genome-environment-association.html#cb567-7" aria-hidden="true" tabindex="-1"></a><span class="co">#fdr level q</span></span>
<span id="cb567-8"><a href="genome-environment-association.html#cb567-8" aria-hidden="true" tabindex="-1"></a>q <span class="ot">=</span> <span class="fl">0.1</span></span>
<span id="cb567-9"><a href="genome-environment-association.html#cb567-9" aria-hidden="true" tabindex="-1"></a>w <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">sort</span>(adj.p.val1) <span class="sc">&lt;</span> q <span class="sc">*</span> (<span class="dv">1</span><span class="sc">:</span>L)<span class="sc">/</span>L)</span>
<span id="cb567-10"><a href="genome-environment-association.html#cb567-10" aria-hidden="true" tabindex="-1"></a><span class="co"># candidates are then</span></span>
<span id="cb567-11"><a href="genome-environment-association.html#cb567-11" aria-hidden="true" tabindex="-1"></a>candidates<span class="fl">.1</span> <span class="ot">=</span> <span class="fu">order</span>(adj.p.val1)[w]</span>
<span id="cb567-12"><a href="genome-environment-association.html#cb567-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb567-13"><a href="genome-environment-association.html#cb567-13" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(candidates<span class="fl">.1</span>)</span>
<span id="cb567-14"><a href="genome-environment-association.html#cb567-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 90</span></span>
<span id="cb567-15"><a href="genome-environment-association.html#cb567-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb567-16"><a href="genome-environment-association.html#cb567-16" aria-hidden="true" tabindex="-1"></a><span class="co">#this is for our first variable Mean Chlorphyl</span></span></code></pre></div>
</div>
<div id="exercises-2" class="section level2" number="12.8">
<h2><span class="header-section-number">12.8</span> Exercises</h2>
<ol style="list-style-type: decimal">
<li>Pick your favorite environmental variable from the bio-oracle website. For marine variables you can use the command <code>list_layers(c("Bio-ORACLE","MARSPEC"))$layer_code</code> to view the different Bio-Oracle and MARSPEC environmental layers. For Terrestrial layers use: <code>list_layers("WorldClim")$layer_code</code> Then download the layer you choose and store it as an object with <code>load_layers()</code>
Now decide which region you want to plot. If you’re interested in getting most of California you can use these coords with the function extent(-125, -115, 30, 40), and then use that to crop your environmental layers and make a plot similiar to the one we made in class.</li>
</ol>
<details>
<summary>
<span style="color: DarkCyan;">Solution</span>
</summary>
<p>
<div class="sourceCode" id="cb568"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb568-1"><a href="genome-environment-association.html#cb568-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sdmpredictors)</span>
<span id="cb568-2"><a href="genome-environment-association.html#cb568-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb568-3"><a href="genome-environment-association.html#cb568-3" aria-hidden="true" tabindex="-1"></a><span class="co">#get env data</span></span>
<span id="cb568-4"><a href="genome-environment-association.html#cb568-4" aria-hidden="true" tabindex="-1"></a>precip_seasonality<span class="ot">&lt;-</span><span class="fu">load_layers</span>(<span class="st">&quot;WC_bio15&quot;</span>)</span>
<span id="cb568-5"><a href="genome-environment-association.html#cb568-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb568-6"><a href="genome-environment-association.html#cb568-6" aria-hidden="true" tabindex="-1"></a>precip_wettest<span class="ot">&lt;-</span><span class="fu">load_layers</span>(<span class="st">&quot;WC_bio13&quot;</span>)</span>
<span id="cb568-7"><a href="genome-environment-association.html#cb568-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb568-8"><a href="genome-environment-association.html#cb568-8" aria-hidden="true" tabindex="-1"></a><span class="co">#make range to crop for california</span></span>
<span id="cb568-9"><a href="genome-environment-association.html#cb568-9" aria-hidden="true" tabindex="-1"></a>cali<span class="ot">&lt;-</span><span class="fu">extent</span>(<span class="sc">-</span><span class="dv">130</span>, <span class="sc">-</span><span class="dv">116</span>, <span class="dv">30</span>, <span class="dv">40</span>)</span>
<span id="cb568-10"><a href="genome-environment-association.html#cb568-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb568-11"><a href="genome-environment-association.html#cb568-11" aria-hidden="true" tabindex="-1"></a><span class="co">#crop it</span></span>
<span id="cb568-12"><a href="genome-environment-association.html#cb568-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb568-13"><a href="genome-environment-association.html#cb568-13" aria-hidden="true" tabindex="-1"></a>precip_w.crop<span class="ot">&lt;-</span><span class="fu">crop</span>(precip_wettest, cali)</span>
<span id="cb568-14"><a href="genome-environment-association.html#cb568-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb568-15"><a href="genome-environment-association.html#cb568-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb568-16"><a href="genome-environment-association.html#cb568-16" aria-hidden="true" tabindex="-1"></a><span class="co">#plot it</span></span>
<span id="cb568-17"><a href="genome-environment-association.html#cb568-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb568-18"><a href="genome-environment-association.html#cb568-18" aria-hidden="true" tabindex="-1"></a>my.colors <span class="ot">=</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;#5E85B8&quot;</span>,<span class="st">&quot;#EDF0C0&quot;</span>,<span class="st">&quot;#C13127&quot;</span>))</span>
<span id="cb568-19"><a href="genome-environment-association.html#cb568-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(precip_w.crop,<span class="at">col=</span><span class="fu">my.colors</span>(<span class="dv">1000</span>),<span class="at">axes=</span><span class="cn">FALSE</span>, <span class="at">box=</span><span class="cn">FALSE</span>)</span>
<span id="cb568-20"><a href="genome-environment-association.html#cb568-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">title</span>(<span class="at">cex.sub =</span> <span class="fl">1.25</span>, <span class="at">sub =</span> <span class="st">&quot;Precipitation of the Wettest Month&quot;</span>)</span></code></pre></div>
<img src="Semester_files/figure-html/unnamed-chunk-114-1.png" width="768" style="display: block; margin: auto;" />
</p>
</details>
<ol start="2" style="list-style-type: decimal">
<li>Use the script for determining the number of candidates for the first environmental variable, find the number of candidates for the other four variables. Are these the same or different SNPs? (hint use merge() or intersect() to find out if the lists of candidates overlap)</li>
</ol>
<details>
<summary>
<span style="color: DarkCyan;">Solution</span>
</summary>
<p>
<div class="sourceCode" id="cb569"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb569-1"><a href="genome-environment-association.html#cb569-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-2"><a href="genome-environment-association.html#cb569-2" aria-hidden="true" tabindex="-1"></a><span class="co">#We&#39;ll use this script below</span></span>
<span id="cb569-3"><a href="genome-environment-association.html#cb569-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-4"><a href="genome-environment-association.html#cb569-4" aria-hidden="true" tabindex="-1"></a><span class="co">#for variable 2 (PH)</span></span>
<span id="cb569-5"><a href="genome-environment-association.html#cb569-5" aria-hidden="true" tabindex="-1"></a><span class="co"># L is number of loci</span></span>
<span id="cb569-6"><a href="genome-environment-association.html#cb569-6" aria-hidden="true" tabindex="-1"></a>L<span class="ot">=</span><span class="dv">3699</span></span>
<span id="cb569-7"><a href="genome-environment-association.html#cb569-7" aria-hidden="true" tabindex="-1"></a><span class="co">#fdr level q</span></span>
<span id="cb569-8"><a href="genome-environment-association.html#cb569-8" aria-hidden="true" tabindex="-1"></a>q <span class="ot">=</span> <span class="fl">0.1</span></span>
<span id="cb569-9"><a href="genome-environment-association.html#cb569-9" aria-hidden="true" tabindex="-1"></a>w <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">sort</span>(adj.p.val2) <span class="sc">&lt;</span> q <span class="sc">*</span> (<span class="dv">1</span><span class="sc">:</span>L)<span class="sc">/</span>L)</span>
<span id="cb569-10"><a href="genome-environment-association.html#cb569-10" aria-hidden="true" tabindex="-1"></a><span class="co"># candidates are then</span></span>
<span id="cb569-11"><a href="genome-environment-association.html#cb569-11" aria-hidden="true" tabindex="-1"></a>candidates<span class="fl">.2</span> <span class="ot">=</span> <span class="fu">order</span>(adj.p.val2)[w]</span>
<span id="cb569-12"><a href="genome-environment-association.html#cb569-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-13"><a href="genome-environment-association.html#cb569-13" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(candidates<span class="fl">.2</span>)</span>
<span id="cb569-14"><a href="genome-environment-association.html#cb569-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 67</span></span>
<span id="cb569-15"><a href="genome-environment-association.html#cb569-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-16"><a href="genome-environment-association.html#cb569-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-17"><a href="genome-environment-association.html#cb569-17" aria-hidden="true" tabindex="-1"></a><span class="co">#variable 3 </span></span>
<span id="cb569-18"><a href="genome-environment-association.html#cb569-18" aria-hidden="true" tabindex="-1"></a>w <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">sort</span>(adj.p.val3) <span class="sc">&lt;</span> q <span class="sc">*</span> (<span class="dv">1</span><span class="sc">:</span>L)<span class="sc">/</span>L)</span>
<span id="cb569-19"><a href="genome-environment-association.html#cb569-19" aria-hidden="true" tabindex="-1"></a><span class="co"># candidates are then</span></span>
<span id="cb569-20"><a href="genome-environment-association.html#cb569-20" aria-hidden="true" tabindex="-1"></a>candidates<span class="fl">.3</span> <span class="ot">=</span> <span class="fu">order</span>(adj.p.val3)[w]</span>
<span id="cb569-21"><a href="genome-environment-association.html#cb569-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-22"><a href="genome-environment-association.html#cb569-22" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(candidates<span class="fl">.3</span>)</span>
<span id="cb569-23"><a href="genome-environment-association.html#cb569-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 78</span></span>
<span id="cb569-24"><a href="genome-environment-association.html#cb569-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-25"><a href="genome-environment-association.html#cb569-25" aria-hidden="true" tabindex="-1"></a><span class="co">#variable 4</span></span>
<span id="cb569-26"><a href="genome-environment-association.html#cb569-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-27"><a href="genome-environment-association.html#cb569-27" aria-hidden="true" tabindex="-1"></a>w <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">sort</span>(adj.p.val4) <span class="sc">&lt;</span> q <span class="sc">*</span> (<span class="dv">1</span><span class="sc">:</span>L)<span class="sc">/</span>L)</span>
<span id="cb569-28"><a href="genome-environment-association.html#cb569-28" aria-hidden="true" tabindex="-1"></a><span class="co"># candidates are then</span></span>
<span id="cb569-29"><a href="genome-environment-association.html#cb569-29" aria-hidden="true" tabindex="-1"></a>candidates<span class="fl">.4</span> <span class="ot">=</span> <span class="fu">order</span>(adj.p.val4)[w]</span>
<span id="cb569-30"><a href="genome-environment-association.html#cb569-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-31"><a href="genome-environment-association.html#cb569-31" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(candidates<span class="fl">.4</span>)</span>
<span id="cb569-32"><a href="genome-environment-association.html#cb569-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 84</span></span>
<span id="cb569-33"><a href="genome-environment-association.html#cb569-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-34"><a href="genome-environment-association.html#cb569-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-35"><a href="genome-environment-association.html#cb569-35" aria-hidden="true" tabindex="-1"></a><span class="co">#variable 5</span></span>
<span id="cb569-36"><a href="genome-environment-association.html#cb569-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-37"><a href="genome-environment-association.html#cb569-37" aria-hidden="true" tabindex="-1"></a>w <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">sort</span>(adj.p.val5) <span class="sc">&lt;</span> q <span class="sc">*</span> (<span class="dv">1</span><span class="sc">:</span>L)<span class="sc">/</span>L)</span>
<span id="cb569-38"><a href="genome-environment-association.html#cb569-38" aria-hidden="true" tabindex="-1"></a><span class="co"># candidates are then</span></span>
<span id="cb569-39"><a href="genome-environment-association.html#cb569-39" aria-hidden="true" tabindex="-1"></a>candidates<span class="fl">.5</span> <span class="ot">=</span> <span class="fu">order</span>(adj.p.val5)[w]</span>
<span id="cb569-40"><a href="genome-environment-association.html#cb569-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-41"><a href="genome-environment-association.html#cb569-41" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(candidates<span class="fl">.5</span>)</span>
<span id="cb569-42"><a href="genome-environment-association.html#cb569-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 113</span></span></code></pre></div>
</p>
</details>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="week-10-genome-wide-association-study-gwas.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Semester.pdf", "Semester.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
